ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Setup/GenChains.hs

```diff
@@ -147,29 +152,59 @@ genChainsWithExtraHonestPeers genNumExtraHonest genNumForks = do
     }
 
   where
-    genAdversarialFragment :: [TestBlock] -> Int -> (Int, [S]) -> AnchoredFragment TestBlock
+    genAdversarialFragment :: [blk] -> Int -> (Int, [S]) -> AnchoredFragment blk
     genAdversarialFragment goodBlocks forkNo (prefixCount, slotsA)
       = mkTestFragment (mkTestBlocks prefix slotsA forkNo)
       where
         -- blocks in the common prefix in reversed order
         prefix = drop (length goodBlocks - prefixCount) goodBlocks
 
-    mkTestFragment :: [TestBlock] -> AnchoredFragment TestBlock
-    mkTestFragment =
-      AF.fromNewestFirst AF.AnchorGenesis
+mkTestFragment :: HasHeader blk => [blk] -> AnchoredFragment blk
+mkTestFragment = AF.fromNewestFirst AF.AnchorGenesis
 
-    mkTestBlocks :: [TestBlock] -> [S] -> Int -> [TestBlock]
-    mkTestBlocks pre active forkNo =
-      fst (List.foldl' folder ([], 0) active)
-      where
-        folder (chain, inc) s | S.test S.notInverted s = (issue inc chain, 0)
-                              | otherwise = (chain, inc + 1)
-        issue inc (h : t) = incSlot inc (successorBlock h) : h : t
-        issue inc [] | [] <- pre = [incSlot inc ((firstBlock (fromIntegral forkNo)) {tbSlot = 0})]
-                     | h : t <- pre = incSlot inc (modifyFork (const (fromIntegral forkNo)) (successorBlock h)) : h : t
-
-    incSlot :: SlotNo -> TestBlock -> TestBlock
-    incSlot n b = b { tbSlot = tbSlot b + n }
+mkTestBlocks :: forall blk. (HasHeader blk, IssueTestBlock blk) => [blk] -> [S] -> Int -> [blk]
+mkTestBlocks pre active forkNo =
+  fst (List.foldl' folder ([], 0) active)
+  where
+    folder :: ([blk], SlotNo) -> S -> ([blk], SlotNo)
+    folder (chain, inc) s | S.test S.notInverted s = (issue inc chain, 0)
+                          | otherwise = (chain, inc + 1)
+    issue :: SlotNo -> [blk] -> [blk]
+    issue inc (h : t) = successorBlock Nothing inc h : h : t
+    issue inc [] =
+      case pre of
+        [] -> [firstBlock forkNo inc]
+        (h : t) -> successorBlock (Just forkNo) inc h : h : t
+
+-- | Class of block types for which we can issue test blocks.
+class IssueTestBlock blk where
```

Comment:
I want to make sure I understand what the fork number described here is and whether we need and can write instances of this class for real blocks, because I don't think they have a fork number.